### PR TITLE
feat(minidump): Support gzipped crashpad requests

### PR DESCRIPTION
**Pull request by jan-auer**
This adds a middleware that decompresses request payloads with a `Content-Encoding` of either `gzip` or `deflate`. We could theoretically also add brotli, in future.

Fixes SDK-296

### CODE REVIEW

**Comment by mattrobenolt** [comment]
Can we not do this as a middleware? Pretty sure this will break /store/ as-is.

**Comment by jan-auer** [comment]
Looks like there was even a test for that. I'll double-check. 

@mattrobenolt this is about store, kind of. At least right now, requests to a store-like endpoint with `Content-Encoding: gzip` result in a 400 response.

**Comment by jan-auer** [comment]
@mattrobenolt https://github.com/getsentry/sentry/blob/fix/minidump-gzip/tests/integration/tests.py#L305-L335 should be covering this.

**Comment by mitsuhiko** [comment]
@jan-auer @mattrobenolt any specific reason we don't want to use the uwsgi supplied transparent decoding here? That would not require us to buffer: https://uwsgi-docs.readthedocs.io/en/latest/Transformations.html

**Comment by mattrobenolt** [comment]
I didn't realize that was a thing, I'd be down for it if we can make it work.

**Comment by jan-auer** [comment]
@mitsuhiko am I missing something or is this only compressing payloads **sent** by uwsgi, and not decompressing those received?

**Comment by mitsuhiko** [comment]
Ugh you might be right. I could swear it has gzip support for requests though.

**Comment by jan-auer** [comment]
I wish, and that's also what [PEP-3333](https://www.python.org/dev/peps/pep-3333/#other-http-features) mentions. But as far as I could see, no dice.

File: src/sentry/middleware/proxy.py [hunk 1]
```diff
@@ -72,6 +77,31 @@ def process_request(self, request):
...
+class DecompressBodyMiddleware(object):
+    def process_request(self, request):
+        encoding = request.META.get('HTTP_CONTENT_ENCODING', '').lower()
+        if encoding in COMPRESSED_ENCODINGS:
+            data = request._stream.read()
+            try:
+                uncompressed = decompress(data, encoding)
+                request._stream = six.StringIO(uncompressed)
+                request.META['CONTENT_LENGTH'] = len(uncompressed)
+                request.META['HTTP_CONTENT_ENCODING'] = 'identity'
```
Comment (jan-auer) [inline src/sentry/middleware/proxy.py]: Wondering if it's better to just remove this here, instead.

Comment (mitsuhiko) [reply src/sentry/middleware/proxy.py]: If we do this at this level I would opt for patching our django to support `wsgi.input_terminated` and have the input stream decode on the fly (there is a gzip file that can do that). We already support reading to the terminator for the chunked request case so it should not be much of a change.

Comment (mitsuhiko) [reply src/sentry/middleware/proxy.py]: That for the record means setting the content length to empty string.

**Review by mitsuhiko — APPROVED** [review]

**Pull request merged**